### PR TITLE
🌎 SDK.capture_html()

### PR DESCRIPTION
**Pull request by asim-shrestha**

### CODE REVIEW

File: sdk/harambe/utils.py [hunk 1]
```diff
@@ -3,8 +3,22 @@
...
 from urllib.parse import urljoin
 
+from nanoid import generate
 from playwright.async_api import Locator, Page
 
+alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+size = 16
+
+
+def nanoid_factory() -> str:
```
Comment (asim-shrestha) [inline sdk/harambe/utils.py:13]: Stolen from deworker for unique file names

File: sdk/harambe/observer.py [hunk 2]
```diff
@@ -32,6 +31,12 @@ class DownloadMeta(TypedDict):
     filename: str
 
 
+class HTMLMetadata(TypedDict):
+    download: DownloadMeta
+    html: str
+    inner_text: str
```
Comment (asim-shrestha) [inline sdk/harambe/observer.py:37]: Thoughts on this API for the sdk function? Can also just have the DownloadMeta fields at the top level (maybe that's better)

Comment (asim-shrestha) [reply sdk/harambe/observer.py]: ended up going with that

File: sdk/test/test_e2e.py [hunk 3]
```diff
@@ -434,3 +434,65 @@ async def scraper(sdk: SDK, current_url: str, *args, **kwargs):
...
+
+
+@pytest.mark.parametrize("harness", [playwright_harness, soup_harness])
+async def test_capture_html_element_not_found(server, observer, harness):
+    url = f"{server}/table"
+
+    @SDK.scraper("test", "detail", observer=observer)
+    async def scraper(sdk: SDK, *args, **kwargs):
+        with pytest.raises(ValueError):
+            await sdk.capture_html("#missing .selector .lies .adam-watkins")
```
Comment (awtkns) [inline sdk/test/test_e2e.py:488]: grimer

File: sdk/harambe/utils.py (hunk 1 above)
Comment (awtkns) [reply sdk/harambe/utils.py:13]: why not just UUID v4 here so we dont need another dep

File: sdk/harambe/core.py [hunk 4]
```diff
@@ -256,6 +259,73 @@ async def capture_download(
...
+        :param exclude_selectors: List of CSS selectors for elements to exclude from capture.
+        :return: HTMLMetadata containing download URL, HTML content and inner text.
+        :raises ValueError: If the specified selector doesn't match any element.
+        """
+        html, inner_text = await self._get_html(selector, exclude_selectors or [])
+
+        with tempfile.NamedTemporaryFile(mode="w", encoding="utf-8") as temp_file:
+            temp_file.write(html)
+            temp_file.flush()
+
```
Comment (awtkns) [inline sdk/harambe/core.py:281]: whats the temp file for? We done seem to use it

File: sdk/pyproject.toml [hunk 5]
```diff
@@ -22,6 +22,7 @@ dependencies = [
     "curl-cffi==0.7.3",
     "ua-generator==1.0.5",
     "python-slugify>=8.0.4",
+    "nanoid>=2.0.0",
```
Comment (awtkns) [inline sdk/pyproject.toml:25]: remove

**Review by awtkns — APPROVED** [review]

**Pull request merged**